services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/schema/JsonSchemaManagerImpl.java

```diff
@@ -549,5 +554,42 @@ public JsonSchemaObjectBinding getJsonSchemaObjectBinding(Long objectId, BoundOb
 	public void clearBoundSchema(Long objectId, BoundObjectType objectType) {
 		jsonSchemaDao.clearBoundSchema(objectId, objectType);
 	}
-
+	
+	@Override
+	public JsonSchema createOrUpdateValidationSchemaIndex(String versionId) {
+		JsonSchemaVersionInfo info = jsonSchemaDao.getVersionInfo(versionId);
+		JsonSchema schema = getValidationSchema(info.get$id());
+		validationIndexDao.createOrUpdate(versionId, schema);
+		// update the schemas that depend on this schema
+		transactionalMessenger.sendMessageAfterCommit(versionId, ObjectType.JSON_SCHEMA_DEPENDANT, ChangeType.UPDATE);
+		// If the semantic version is null, notify the entities bound to it, because
+		// schemas with no semantic version are mutable.
+		if (info.getSemanticVersion() == null) {
+			sendUpdateNotifications(info.getSchemaId());
+		}
+		return schema;
+	}
+	
+	@Override
+	public void sendValidationIndexNotificationsForDependants(String versionId) {
+		// case where versionId is a schema with semantic version
+		JsonSchemaVersionInfo info = jsonSchemaDao.getVersionInfo(versionId);
+		Iterator<String> versionIds = null;
+		// semantic version is non-null, then versionId is non-null in dependency table
+		if (info.getSemanticVersion() != null) {
+			versionIds = jsonSchemaDao.getVersionIdsOfDependantsIteratorWithVersionId(versionId);
+		} else {
+			// otherwise grab the versionId's schemaId from the latest version table to access the dependency table
+			String schemaId = info.getSchemaId();
+			versionIds = jsonSchemaDao.getVersionIdsOfDependantsIteratorWithSchemaId(schemaId);
+		}
+		while (versionIds.hasNext()) {
+			transactionalMessenger.sendMessageAfterCommit(versionIds.next(), ObjectType.JSON_SCHEMA, ChangeType.UPDATE);
```

Comment:
if the iterator is recursive it will send JSON_SCHEMA events for the full dependency tree.  Let the Database remove duplicates.  The query would use the WITH RECURSIVE feature (https://dev.mysql.com/doc/refman/8.0/en/with.html#common-table-expressions-recursive-examples)
Also write the SQL as a resource and load using :
	public static final String ENTITY_DEPTH_SQL = DDLUtilsImpl
			.loadSQLFromClasspath("sql/EntityDepth.sql");